feat: redesign options for the `quotes` rule

snitin315 opened the pull request:
<!--
    Thank you for contributing!

    ESLint adheres to the [JS Foundation Code of Conduct](https://js.foundation/community/code-of-conduct).
-->

#### Prerequisites checklist

- [X] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

<!--
    The following template is intentionally not a markdown checkbox list for the reasons
    explained in https://github.com/eslint/eslint/pull/12848#issuecomment-580302888
-->

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[X] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

<!--
    If the item you've checked above has a template, please paste the template questions below and answer them. (If this pull request is addressing an issue, you can just paste a link to the issue here instead.)
-->

<!--
    Please ensure your pull request is ready:

    - Read the pull request guide (https://eslint.org/docs/developer-guide/contributing/pull-requests)
    - Include tests for this change
    - Update documentation for this change (if appropriate)
-->

<!--
    The following is required for all pull requests:
-->

#### What changes did you make? (Give an overview)

Fix https://github.com/eslint/eslint/issues/12156

Redesign options for the `quotes` rule as per https://github.com/eslint/eslint/issues/12156#issuecomment-601512762

`/* eslint quotes: ['error', { precedence: ['single', 'double', 'backtick'], avoidEscape: true } ]*/`

#### Is there anything you'd like reviewers to focus on?

aladdin-add commented:
are the old-style options allowed after the change?  I think they can be auto translated to the new-style options -- to avoid breaking changes.

snitin315 commented:
I was considering this as a candidiate for v9 because of the breaking changes.

But, yes, we should keep the old-style schema for avoiding breaking changes if we want to ship this feature with v8.

And then we can completely remove the old syntax in v9.

nzakas commented:
@snitin315 are you still working on this? There’s linting and test errors.

Also, breaking changes need to be flagged as “feat!:”, and need to be discussed separately. It’s a bit early to start sending PRs for v9 :)

snitin315 commented:
Should we keep the old syntax as well to avoid breaking changes? So that we can ship this new syntax with v8. We could log a deprecation warning when when using the old syntax.

aladdin-add commented:
For a smoother UX, I would prefer to also allow the old options if it does not add too many maintenances burden.
and, we don't have a rule options deprecation process (though good to have IMO :)).

nzakas commented:
Yeah, we don’t really have a way to deprecate old config formats. We need to continue supporting the old way going forward, just mark it as deprecated in the docs.

noahsark769 commented:
This is awesome! @snitin315 happy to help with this if I can, also agreed about keeping the old syntax compatible

snitin315 commented:
This feature is definitely going to be a big breaking change. Even if we allow the old syntax by converting the old syntax options to new syntax internally, the working of this rule will be changed and it will produce more warnings.

### Consider the following example with old syntax - 

```js
/*eslint quotes: ["error", "double", { "allowTemplateLiterals": true }]*/

let double = "double"; // ok
double = `double`; // ok
double = `has ' single`; // ok
double = `has " double`; // ok
```

### Possible conversion to new syntax as per https://github.com/eslint/eslint/issues/12156#issuecomment-601512762 - 

```js
/*eslint quotes: ["error", { precedence: ["double", "backtick"] }]*/

let double = "double"; // ok
double = `double`; // error,  fixed to "double". 
double = `has ' single`; // error,  fixed to "has ' single". 
double = `has " double`; // error,  fixed to "has \" double". 
```

```js
/*eslint quotes: ["error", { precedence: ["double", "backtick"], avoidEscape: true }]*/

let double = "double"; // ok
double = `double`; // fixed to "double".
double = `has ' single`; // error,  fixed to "has ' single". 
double = `has " double`; // ok
```

snitin315 commented:
I am closing this PR for now since it's too early for v9 PRs. I will reopen when we are close to v9.
